[User History] migrate to structured messages & add for domain

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/USH-1196

:fish: 

This PR updates the change messages to a new structured format
[Initial discussion](https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit) and then finally decision to [go with translations approach](https://github.com/dimagi/commcare-hq/pull/30181#discussion_r683929764) which was further simplified
```
Each change message to follow the structure
    {
        "field": {
            "slug": params, # params for the message
        }
    }
    field: could be domain, phone_numbers etc as the top-level key to make them searchable
    slug: any slug in MESSAGES
    params: (optional) params for the message
```

Working on a rollout plan in https://github.com/dimagi/commcare-hq/pull/30287/files though it should be okay to deploy this without waiting on it since the old columns are not removed here and will be removed post migration only.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`USER_HISTORY_REPORT`

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added new tests and updated tests for user importer for relevant changes.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3465

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Added tests for changes.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

This PR puts logging into a new structure, hence revert should not be done unless it's unfixable. And also, it should be checked if the old columns have been removed or not (migration PR linked above)

mkangia committed:
add new top level columns

- top level makes its possible to search for fields changed
- extract changed_via too and remove details column entirely

mkangia committed:
fix documentation

mkangia committed:
Update change message to use message slugs

mkangia committed:
pass correct argument

- update some simpler message dict creation

mkangia committed:
pass locations

- use the new location change methods

mkangia committed:
fix args: pass profile id now

mkangia committed:
fix arg: pass groups now

mkangia committed:
use new bulk phone numbers method

mkangia committed:
use new fields in tests

- detail['changes'] -> changes
- detail['changed_via'] -> changed_via
- message -> change_messages

mkangia committed:
update report

- use new changes column
- use new changed_via
- use get_messages

mkangia committed:
Do ToDo

mkangia committed:
pass list since set not serializable

mkangia committed:
fix tests missed earlier

- in commit "use new fields in tests"

mkangia committed:
fix some simple tests

mkangia committed:
support multiple changes per field

mkangia committed:
update bulk upload tests

mkangia committed:
Error if duplicate slug entry

mkangia committed:
combine two factor disable change messages

mkangia committed:
fix test

Failure in case the order of phone numbers in params phone_numbers is
different for multiple numbers

mkangia committed:
refactor & add comment

- user.get_group_ids calls Group.by_user_id internally, so just call that directly when no group ids passed

snopoke commented on corehq/apps/users/audit/change_messages.py:
```diff
@@ -7,8 +7,7 @@ class UserChangeMessage(object):
         "field": {
-            "slug": message_slug,
-            "params": {} # optional params for the message
+            "slug": params, # params for the message
```
:+1:

snopoke commented on corehq/apps/user_importer/tests/test_importer.py:
```diff
@@ -1727,3 +1729,98 @@ def test_invite_location_add(self):
+        # no exception raised for new user
+        user_change_logger.add_change_message(UserChangeMessage.password_reset())
+
+        user = CommCareUser.create(self.domain_name, f"hello@{self.domain.name}.commcarehq.org", "*******",
+                                   created_by=None, created_via=None)
+        self.addCleanup(user.delete, self.domain_name, deleted_by=None)
```
do actually need to save the user here? I think you can just pass in `is_new_user=False` with an unsaved user model can't you? Perhaps then you could make this a non-db test.

snopoke commented on corehq/apps/user_importer/tests/test_importer.py:
```diff
@@ -1727,3 +1729,98 @@ def test_invite_location_add(self):
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.domain_name = 'mydomain'
+        cls.domain = Domain.get_or_create_with_name(name=cls.domain_name)
+        cls.uploading_user = WebUser.create(cls.domain_name, "admin@xyz.com", 'password', None, None,
+                                            is_superuser=True)
+        cls.upload_record = UserUploadRecord(
+            domain=cls.domain_name,
+            user_id=cls.uploading_user.get_id
+        )
+        cls.upload_record.save()
```
is it really necessary to create these objects for this test? I think you only need 'real' values if you call `user_change_logger.save()` which you aren't doing here.

snopoke commented on corehq/apps/hqadmin/views/users.py:
```diff
@@ -480,29 +480,27 @@ def get(self, request, *args, **kwargs):
         for device in devices_for_user(user):
             device.delete()
-        change_messages.update(UserChangeMessage.registered_devices_reset())
+            devices_reset = True
```
do we need to record this? It isn't an option action so we could assume that if 2fa was disabled then it was done right? I don't see any additional value in recording it.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/30253/commits/48cb8fad73224f5f978040fb32736e5a3d99301d

snopoke commented on corehq/apps/users/audit/change_messages.py:
```diff
@@ -54,33 +54,21 @@ def domain_removal(domain):
             }
         }
+        if devices_reset:
+            change_message["two_factor"].update({"reset_devices": {}})
```
as mentioned above I don't think there's any value in recording this. Resetting the devices is the mechanism by which we reset 2fa.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/30253/commits/48cb8fad73224f5f978040fb32736e5a3d99301d

snopoke commented on corehq/apps/users/tests/test_audit.py:
```diff
@@ -110,46 +112,59 @@ def test_add_as_web_user(self):
             },
-            'Two factor disabled. Verified by: jamesbond@mi6.com, verification mode: "007"'
+            [
+                'Two factor disabled. Verified by: jamesbond@mi6.com, verification mode: "007"',
```
I don't think this message should say `disabled` but rather `removed`.

mkangia committed:
use models; since not saving

mkangia committed:
avoid saving objects

mkangia committed:
remove unused import

mkangia committed:
disabled -> removed

mkangia committed:
extract slugs as constants

mkangia committed:
add for_domain column

- rename domain -> by_domain

mkangia committed:
pass for_domain

mkangia committed:
update tests

mkangia committed:
remove unused method

esoergel reviewed (approved):
Didn't do a super thorough review, but the general approach looks good to me.

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -209,24 +209,24 @@ def update_user(self, metadata_updated=False, profile_updated=False):
 
         if is_update_successful and (props_updated or role_updated or metadata_updated):
-            messages = []
+            change_messages = {}
```
There is actually [UserChangeLogger](https://github.com/dimagi/commcare-hq/blob/5fdfc6570d08acee9ec0a7ab19a8a6d43c377173/corehq/apps/user_importer/helpers.py#L21) that does that and should be used if we need to assemble multiple messages.
If we still need something to assemble messages, it would be better to add a new class for it, so that `UserChangeMessage` can just focus on making messages as its responsibility, though I think `UserChangeLogger` does that well by also adding messages [only when needed](https://github.com/dimagi/commcare-hq/blob/5fdfc6570d08acee9ec0a7ab19a8a6d43c377173/corehq/apps/user_importer/helpers.py#L53) and also [check on duplicates](https://github.com/dimagi/commcare-hq/blob/5fdfc6570d08acee9ec0a7ab19a8a6d43c377173/corehq/apps/user_importer/helpers.py#L71).

though it currently sits in `user_importer` app because that is where it was originally added but can definitely be moved to `users` app.

mkangia committed:
use constants

mkangia commented on corehq/apps/users/audit/change_messages.py:
```diff
@@ -1,91 +1,314 @@
+        if devices_reset:
+            change_message["two_factor"].update({"reset_devices": {}})
+        if disable_for_days:
+            change_message["two_factor"].update({"disable_for_days": {"days": disable_for_days}})
```
https://github.com/dimagi/commcare-hq/pull/30253/commits/0459ff4c6f52f112fd71d0f0fc66a620d644e394

mkangia committed:
make for_domain mandatory as well

mkangia committed:
Merge branch 'master' into mk/user-history/structured-messages-v2

mkangia committed:
merge mk/user-history/structured-messages-v2; resolve conflict

mkangia committed:
skip tracking devices reset

mkangia committed:
merge mk/user-history/structured-messages-v2; resolve conflict

mkangia committed:
keep requirement separate

-> domain_required_for_log -> by_domain_required_for_log

mkangia committed:
fix user delete

mkangia committed:
log when locations are removed for commcare user

we already log these under "changes" for mobile users, but to keep consistency with web users log change messages as well

mkangia committed:
Merge branch 'mk/user-history/structured-messages-v2' into mk/user-history/add-for-domain

mkangia committed:
fix report; use new name for column

mkangia committed:
update test for additional logging

mkangia committed:
Merge branch 'master' into mk/user-history/structured-messages-v2

mkangia committed:
update to use new methods

mkangia committed:
Merge branch 'mk/user-history/structured-messages-v2' into mk/user-history/add-for-domain

mkangia committed:
update to use new args for api logging

mkangia committed:
update to use new args for new phone removal logging

mkangia commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -231,6 +231,8 @@ def update_locations(self, location_codes, domain_info):
                 self.logger.add_info(
                     UserChangeMessage.assigned_locations_info(locations))
+            else:
+                self.logger.add_info(UserChangeMessage.assigned_locations_info([]))
```
migration for old records to be handled in https://github.com/dimagi/commcare-hq/pull/30287

mkangia committed:
avoids errors by adding phone number changes in one call

mkangia committed:
fixes groups log

mkangia committed:
fixes test with new fields

mkangia committed:
Merge branch 'mk/user-history/structured-messages-v2' into mk/user-history/add-for-domain

mkangia committed:
Merge pull request #30301 from dimagi/mk/user-history/add-for-domain

[User History] Add for domain

mkangia committed:
fix tests to use new columns

mkangia commented on corehq/apps/users/util.py:
```diff
@@ -314,60 +314,56 @@ def _last_sync_needs_update(last_sync, sync_datetime):
     :param fields_changed: dict of user fields that have changed with their current value
     :param action: action on the user
-    :param domain_required_for_log: set to False to allow domain less log for specific changes
+    :param by_domain_required_for_log: set to False to allow domain less log for specific changes
+    :param for_domain_required_for_log: set to False to allow domain less log for specific changes
```
It is made to be the caller's responsibility to be explicit that it is okay for the log to be without to_ or for_ domain.
The function itself can't really decide for that because it depends on where the logging is happening in the code.

Does that make sense?

esoergel replied:
Yeah, that makes sense, but then why have the check in `log_user_change` at all?  If it's up to the caller to decide whether `for_domain` is appropriate in this context, couldn't the caller just pass in the domain when available and `None` when it isn't?  Are you thinking about scenarios where the caller might have a variable called `domain` but that's actually set to `None` for some reason?

mkangia replied:
The check is to ensure that for_domain is passed.

>  Are you thinking about scenarios where the caller might have a variable called domain but that's actually set to None for some reason?

Yes, something like that where domain could be blank unknowingly. Like places where we are using `request.domain` or other ways to get domain, just a safety net to ensure domain is always passed unless explicitly marked as not required.

mkangia committed:
merge master

mkangia committed:
add help link

mkangia committed:
simplifies by getting domain from request itself

mkangia committed:
uses cleaner message

mkangia committed:
uses info for info kind messages

this does not change what gets stored in log

mkangia committed:
tracks full domain transfer

mkangia committed:
displays all columns

- remove unnecessary action filter to avoid a query

mkangia commented:
Here it goes!
